### ISSUE TITLE
trial's "test_suite" feature violates naming standard

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2165]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2165|
|Type|defect|
|Created|2006-10-10 14:29:25Z|

This should be "testSuite" or something similar.  Since it's a factory, I would expect the name to give some indication of this as well.

<details><summary>Searchable metadata</summary>

```
trac-id__2165 2165
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1160490565000000 1160490565000000
changetime__1163895622000000 1163895622000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

You mean the method used for a module to specify its own suite? That's called `test_suite` because there seems to be an emerging convention across frameworks. See [#7250](https://github.com/twisted/twisted/issues/7250)

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_2"></a>|@radix commented|
|-|-|

I suggest supporting both testSuite and test_suite. And please let's get this resolved by the time 2.5 is released.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to @radix|
|-|-|

Ready for review in testSuite-2165

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @jml|
|-|-|

One existing problem and one new problem: the way test_suite and testSuite are invoked, if either raises an AttributeError, it will be swallowed.

mockcustomsuite files should have copyright headers and at least a module docstring explaining their use.

Thanks for the whitespace cleanup, and good catch on the case where both testSuite and test_suite are defined by a single module.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml set owner to @exarkun|
|-|-|

It is done. Ready for re-review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @jml|
|-|-|

Hee hee heeee heeeee.  I didn't expect this to come up so soon:

```
$ trial twisted
/home/exarkun/Projects/Twisted/trunk/twisted/test/test_persisted.py:161: DeprecationWarning: twisted.persisted.marmalade is deprecated
  from twisted.persisted import marmalade
/home/exarkun/Projects/Twisted/trunk/twisted/words/im/__init__.py:8: UserWarning: twisted.im will be undergoing a rewrite at some point in the future.
  warnings.warn("twisted.im will be undergoing a rewrite at some point in the future.")
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/bin/trial", line 24, in ?
    run()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/scripts/trial.py", line 337, in run
    suite = _getSuite(config)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/scripts/trial.py", line 297, in _getSuite
    return loader.loadByNames(config['tests'], recurse)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 568, in loadByNames
    suites = [self.loadAnything(thing, recurse)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 526, in loadAnything
    return self.loadPackage(thing, recurse)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 493, in loadPackage
    thingToAdd = self.loadModule(module)
  File "/home/exarkun/Projects/Twisted/trunk/twisted/trial/runner.py", line 405, in loadModule
    return module.testSuite()
  File "/home/exarkun/Projects/Twisted/trunk/twisted/words/test/test_toc.py", line 343, in testSuite
    def testSuite():return unittest.TestSuite(map(lambda x:x(),testCases))
AttributeError: 'module' object has no attribute 'TestSuite'
exarkun@boson:/tmp$ 
```

A bug in test_toc, I would say.  We can just delete that testSuite method, right?

Other than that, the branch looks good.  One minor point I'd like to clarify, though.  I think the initial approach of catching AttributeError was the correct one, there was just too much code in the try suite.  Something like this:

```python
try:
    x.y()
except AttributeError:
    pass
```

Can be rewritten like this:

```python
try:
    y = x.y
except AttributeError:
    pass
else:
    y()
```

to avoid accidentally catching too many exceptions.  This is better than hasattr, because
hasattr will often silently swallow exceptions as well. :/  For the specific case in question
here, it _probably_ makes no difference which of these is used, because the object the attribute is being loaded from is a module, and modules don't or can't have custom descriptors, so the hasattr error case should never come up.  I just wanted to point this out the distinction in case something similar comes up in the future.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml set status to `closed`|
|-|-|

From r18718:

Provide `testSuite` as a synonym for `test_suite`
 * Author: jml
 * Reviewer: exarkun
 * Fixes: [#2165](https://github.com/twisted/twisted/issues/2165)

`test_suite` allows test authors to provide a their own test suite
for a module. Now, `testSuite` does the same. If they are both present
only `testSuite` will be consulted.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
